types/underscore/index.d.ts

```diff
@@ -813,54 +815,40 @@ declare module _ {
-        * contains all of the second elements, and so on. Use with apply to pass in an array
-        * of arrays
-        * @param arrays The arrays to unzip.
-        * @return Unzipped version of `arrays`.
-        **/
-        unzip(...arrays: any[][]): any[][];
-
-        /**
-        * Converts arrays into objects. Pass either a single list of [key, value] pairs, or a
-        * list of keys, and a list of values.
-        * @param keys Key array.
-        * @param values Value array.
-        * @return An object containing the `keys` as properties and `values` as the property values.
-        **/
-        object<TResult extends {}>(
-            keys: _.List<string>,
-            values: _.List<any>): TResult;
+         * Merges together the values of each of the `lists` with the values at
+         * the corresponding position. Useful when you have separate data
+         * sources that are coordinated through matching list indexes. If
+         * you're working with a matrix of nested lists, this can be used to
+         * transpose the matrix.
+         * @param lists The lists to zip.
+         * @returns The zipped version of `lists`.
+         **/
+        zip(...lists: List<any>[]): any[][];
 
         /**
-        * Converts arrays into objects. Pass either a single list of [key, value] pairs, or a
-        * list of keys, and a list of values.
-        * @param keyValuePairs Array of [key, value] pairs.
-        * @return An object containing the `keys` as properties and `values` as the property values.
-        **/
-        object<TResult extends {}>(...keyValuePairs: any[][]): TResult;
+         * The opposite of zip. Given a list of lists, returns a series of new
+         * arrays, the first of which contains all of the first elements in the
+         * input lists, the second of which contains all of the second
+         * elements, and so on.
+         * @param lists The lists to unzip.
+         * @returns The unzipped version of `lists`.
+         **/
+        unzip(lists: List<List<any>>): any[][];
 
         /**
-        * @see _.object
-        **/
-        object<TResult extends {}>(
-            list: _.List<any>,
-            values?: any): TResult;
+         * Converts lists into objects. Pass either a single `list` of
+         * [key, value] pairs, or a `list` of keys and a list of `values`.
+         * Passing by pairs is the reverse of pairs. If duplicate keys exist,
+         * the last value wins.
+         * @param list A list of [key, value] pairs or a list of keys.
+         * @param values If `list` is a list of keys, a list of values
+         * corresponding to those keys.
+         * @retursn An object comprised of the provided keys and values.
+         **/
+        object<V extends List<any>, TValue extends PairValueOrAny<TypeOfList<V>> = PairValueOrAny<TypeOfList<V>>>(
+            list: V,
+            values?: List<TValue>
+        ): Dictionary<TValue>;
```

Comment:
Something that I considered accounting for here but that I'm not sure whether it's worth it or not is that passing a list of keys that is longer than the list of values produces an object that contains some `undefined` values. How do folks feel about leaving `undefined` out of the possible results for the dictionary values?

If I put it back, ideally it should only apply to the case where `list` is keys rather than tuples since tuples don't allow for this possibility, which might make it tricky to keep this as a single overload.

Reply:
Yes, I think this is a case where having two overloads is warranted, not only because of the potential `undefined` properties but also because `List<any>` seems overly liberal here. `V` can be either `List<EnumerableKey>` or `List<[EnumerableKey, T]>` (with `T` unconstrained). That tuple type might be too restrictive if there are cases where TS can't infer it, but maybe you could work around that with something like `{'0': EnumerableKey, '1': T} | List<EnumerableKey | T>` or otherwise even `List<any>` (so the overall `V` becomes `List<List<any>>`).

Reply:
In fact, maybe three overloads: one for the two-argument case, one for the one-argument case where the element type can be inferred, and one for the one-argument case where the element type falls back to `any`.

Reply:
Good news! I think I found an admittedly mildly complicated way to do this that addresses the issues of `List<any>` being too liberal and `undefined` not being included in the result and have added or adjusted tests as appropriate to reflect that. Let me know what you think.